[fdk-aac] make patent-encumbered HE-AAC optional

Be-ing opened the pull request:
By default, use a fork of fdk-aac (which was originally written
by Fraunhofer for the Android Open Source Project) which has
support for the patent-encumbered HE-AAC, HE-AACv2, and xHE-AAC
profiles removed. This fork is used by Fedora and Arch Linux to
provide support for most AAC use cases without patent licensing
fees and to permit combining it with GPL software. The upstream
fdk-aac with support for all AAC profiles can be built with the
new 'he-aac' option for this port.

Fedora Legal's opinion, from
https://bugzilla.redhat.com/show_bug.cgi?id=1501522#c112 :
The Fedora Project is aware that the Free Software Foundation
has stated that the Fraunhofer FDK AAC license is GPL
incompatible, specifically, because of Clause 3.

We believe that the fdk-aac software codec implementation that we
wish to include in Fedora is no longer encumbered by AAC patents.
This fact means that Clause 3 in the FDK AAC license is a "no op",
or to put it plainly, if no patents are in play, there are no
patent licenses to disclaim. For this (and only this) specific
implementation of fdk-aac, we believe that the FDK AAC license is
GPL compatible.

Be-ing commented:
I have opened a [merge request](https://gitlab.freedesktop.org/wtaymans/fdk-aac-stripped/-/merge_requests/1) to add CMake support to the fork without HE-AAC.

Be-ing commented:
I have [asked upstream to tag a new release](https://github.com/mstorsjo/fdk-aac/issues/130). Let's wait a bit for them to respond before merging this.

Be-ing commented:
@Sibras I am wondering if this PR might present issues building ffmpeg with the fdk-aac option. Should that build fdk-aac with the new he-aac option introduced in this PR?

Sibras commented:
> 
> 
> @Sibras I am wondering if this PR might present issues building ffmpeg with the fdk-aac option. Should that build fdk-aac with the new he-aac option introduced in this PR?

Yes this may cause some issues. Since this PR changes the default behavior of fdk perhaps make the he-aac option default to on (so optionally disable) to keep that libs functionality consistent.

Also the ffmpeg project doesn't recognize the removal of HE-AAC parts as making it GPL compatible so no matter if its included or not ffmpeg with fdk still requires the non-redistributable license.

Be-ing commented:
ffmpeg is the only port in vcpkg that depends on fdk-aac, so I think it would be sufficient to change the ffmpeg port to depend on `fdk-aac[he-aac]` instead of just `fdk-aac` for ffmpeg's `fdk-aac` option.

Be-ing commented:
> perhaps make the he-aac option default to on

Let's default to not putting users in a questionable legal situation.

Be-ing commented:
From https://bugzilla.redhat.com/show_bug.cgi?id=1501522#c29 :
> Christian, supporting HE-AAC encoding is the raison d'être of fdk-aac. Otherwise FFmpeg's aac codec is good enough for all other use cases. We already have gstreamer1-libav in RPMFusion covering these, so having a crippled fdk-aac in Fedora is no gain. See https://trac.ffmpeg.org/wiki/Encode/AAC for more information. Unless we're looking at several years until unstripped fdk-aac can be included in Fedora, I think it's not worth the effort to add support for this stripped version to FFmpeg and other multimedia software and we should just wait.

So there wouldn't be much of a point to building ffmpeg with fdk-aac using the fdk-aac fork without HE-AAC. In Mixxx we have already implemented encoding using fdk-aac, so we'll be using the fork without HE-AAC in our Windows and macOS packages.

Be-ing committed:
[fdk-aac] make patent-encumbered HE-AAC optional

By default, use a fork of fdk-aac (which was originally written
by Fraunhofer for the Android Open Source Project) which has
support for the patent-encumbered HE-AAC, HE-AACv2, and xHE-AAC
profiles removed. This fork is used by Fedora and Arch Linux to
provide support for most AAC use cases without patent licensing
fees and to permit combining it with GPL software. The upstream
fdk-aac with support for all AAC profiles can be built with the
new 'he-aac' option for this port.

Fedora Legal's opinion, from
https://bugzilla.redhat.com/show_bug.cgi?id=1501522#c112 :
The Fedora Project is aware that the Free Software Foundation
has stated that the Fraunhofer FDK AAC license is GPL
incompatible, specifically, because of Clause 3.

We believe that the fdk-aac software codec implementation that we
wish to include in Fedora is no longer encumbered by AAC patents.
This fact means that Clause 3 in the FDK AAC license is a "no op",
or to put it plainly, if no patents are in play, there are no
patent licenses to disclaim. For this (and only this) specific
implementation of fdk-aac, we believe that the FDK AAC license is
GPL compatible.

Also:
* remove restriction on dynamic linking. Upstream has a .def
file which is used by CMake.
* replace deprecated vcpkg functions

Be-ing commented:
Rebased on #18549.

Be-ing committed:
[ffmpeg] use fdk-aac[he-aac] for fdk-aac feature

There is no point to this feature without HE-AAC because FFmpeg has
its own AAC codec that does not support HE-AAC.

Be-ing commented:
I am confused why CI is failing. Running `./vcpkg format-manifest ports/fdk-aac/vcpkg.json` succeeds...

Be-ing commented on ports/ffmpeg/vcpkg.json:
```diff
@@ -299,9 +299,15 @@
-        "fdk-aac",
+        {
+          "name": "fdk-aac",
+          "default-features": false,
```
I am unsure if this line is necessary.

NancyLi1013 replied:
@mcmtroffaes Could you please help review this point? Thanks.

mcmtroffaes replied:
Yes, ``"default-features": false`` is necessary. Generally you don't want features to pull in the full set of default features just for a single codec.

NancyLi1013 replied:
Thanks for your help @mcmtroffaes.

Be-ing commented:
@Sibras could you take a look at this?

Be-ing commented:
> I am confused why CI is failing. Running ./vcpkg format-manifest ports/fdk-aac/vcpkg.json locally succeeds...

Hah, I was running format-manifest on the wrong port. I had to switch to using the dependencies JSON object to use the he-aac feature of the fdk-aac port, which was surprising because I expected `fdk-aac[he-aac]` to work like it does with the CLI.

mcmtroffaes commented on ports/fdk-aac/vcpkg.json:
```diff
@@ -1,6 +1,22 @@
-  "homepage": "https://github.com/mstorsjo/fdk-aac"
+  "port-version": 1,
+  "description": "A standalone library of the Fraunhofer FDK AAC code from the Android Open Source Project. Uses a fork without HE-AAC, HE-AACv2, or xHE-AAC support to avoid patent licensing and GPL compatibility issues when built without the he-aac option.",
+  "homepage": "https://github.com/mstorsjo/fdk-aac",
```
I wonder if this should this point to the stripped version instead (since that's what's installed by default, without the he-aac feature)? I guess there's no way to include two homepages...

Be-ing replied:
:shrug:

mcmtroffaes commented:
LGTM!

NancyLi1013 commented on ports/fdk-aac/portfile.cmake:
```diff
@@ -1,21 +1,34 @@
+vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/${PORT})
 vcpkg_fixup_pkgconfig()
 vcpkg_copy_pdbs()
 vcpkg_copy_tool_dependencies(${CURRENT_PACKAGES_DIR}/tools/${PORT})
```
```suggestion
vcpkg_copy_tool_dependencies("${CURRENT_PACKAGES_DIR}/tools/${PORT}")
```
`file(INSTALL "${SOURCE_PATH}/NOTICE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)`

Please add double quotes to the paths.

NancyLi1013 replied:
Seems you forgot to update these two lines.

NancyLi1013 commented:
Have you tested the feature `he-aac` on your local? @Be-ing

Be-ing commented:
Yes, this port works both with and without the he-aac feature.

Be-ing commented:
Merge?

NancyLi1013 commented:
Waiting for merge now. @Be-ing

BillyONeal commented on ports/fdk-aac/vcpkg.json:
```diff
@@ -1,6 +1,22 @@
-  "description": "A standalone library of the Fraunhofer FDK AAC code",
-  "homepage": "https://github.com/mstorsjo/fdk-aac"
+  "port-version": 1,
+  "description": "A standalone Third-Party Modified Version of the Fraunhofer FDK AAC Codec Library for Android. Uses a fork without HE-AAC, HE-AACv2, or xHE-AAC support to avoid patent licensing and GPL compatibility issues when built without the he-aac option.",
```
```suggestion
  "description": "The Fraunhofer FDK AAC Codec Library. By default, uses a fork without HE-AAC, HE-AACv2, or xHE-AAC support to avoid patent licensing and GPL compatibility issues. For the original, un-forked version, use the \"he-aac\" feature.",
```

Be-ing replied:
This is required by the license: https://gitlab.freedesktop.org/wtaymans/fdk-aac-stripped/-/blob/stripped4/NOTICE#L54

BillyONeal replied:
Hmmmm.... the "for Android" part is the really unfortunate part. I do want this to say something to the effect of "by default uses the fork" rather than always saying it uses the fork.

Be-ing replied:
I thought that was clear already but if you have a suggestion for better wording please share.

BillyONeal commented:
I'm a bit concerned about changing what we consider an authoritative source for an already shipping port; under normal circumstances we would require adding a new port (for example, `fdk-aac-stripped`) instead. However, because this is supposedly the same library with a feature removed, it might be OK. How often does upstream fdk-aac change and are you going to be on the hook to apply such changes in a timely manner?

ras0219-msft commented:
It looks like this is at least supported by the fedora project: https://src.fedoraproject.org/rpms/fdk-aac-free

BillyONeal commented:
> It looks like this is at least supported by the fedora project: https://src.fedoraproject.org/rpms/fdk-aac-free

@Be-ing Can you confirm that this is expected to be "supported" tracking upstream closely and apply my suggested change to the description? (I tried to apply that for you but this PR was created without the "allow maintainers to make edits" option checked) Then we'll merge.

Be-ing commented:
Yes, this will be maintained. [Alpine Linux is considering to package the fork too](https://gitlab.alpinelinux.org/alpine/aports/-/issues/12851). Now that I [fixed the merge conflicts and CMake build in the fork](https://gitlab.freedesktop.org/wtaymans/fdk-aac-stripped/-/merge_requests/1), maintaining the fork should be a simple matter of merging upstream when there is a release.

BillyONeal commented:
Thanks for your contribution!
